### ISSUE TITLE
Links using FB in twin (console)

### ISSUE THREAD

**Issue by mikedebian** [cosmos72/twin#24]
I don't know/think if it's possible to run links (web browser) in graphical mode under twin in console mode.

links -g bugs out with

```
Could not initialize any graphics driver. Tried the following drivers:
x:
Can't open display "(null)"
fb:
Could not get VT mode.
```

It would be a great feature to run links in graphical mode under twin with no X. 
However, I understand if it's not feasible. Just an idea.

I would like to add that I can run a 1440p video under twin using mplayer -vo fbdev2 with no issues.

mplayer -vo fbdev2 -x 1280 -y 720 File.extension works perfectly, as an example.

I understand this might be a limitation with Links, but I'd love to have a working graphical web browser in twin. I think it'd compliment it.

**Comment by rodneyrod** [comment]
Does grapical mode in Links use the framebuffer or does it render using the terminal?

**Comment by Mirppc** [comment]
Umm, Links has no graphical mode as it is all Ncurses and terminal based.  Tested and Links works for me.

**Comment by mikedebian** [comment]
man links

```
-driver <driver name>
              Graphics driver to use. Drivers are: x, svgalib, fb, directfb, pmshell, atheos.  List of drivers will be shown if you give it an unknown driver.  Available drivers depend on your operating system and available libraries.
              Pass a parameter to the graphics driver. On SVGAlib, this is widthxheightxdepth, (where depth is 16, 256, 32K, 64K, 16M or 16M32) and defines the graphics mode, on X the format is WIDTHxHEIGHT and defines the window size
```

Try running links -g in console.If your permissions are set up correctly, it will use the framebuffer to display color and art.

**Comment by cosmos72** [comment]
Hello @mikedebian,

as you pointed out, links -g tries to start links in graphical mode.
In your case, it failed to use both X and framebuffer.

It's easy to imagine that you did not have X running during your test, while I do not know why it failed to use the framebuffer - maybe links wants to be launched from an actual console /dev/tty[0-6][0-9] instead of a pseudo-tty /dev/pts/* that twin uses to run terminal-based programs ?

I can think of several troubleshooting ideas - the simplest are:
* run "links -g" as root inside twin
* run "strace links -g" and analyze strace output

Also, it would be nice if you can confirm that "links -no-g" i.e. in pure text mode works fine inside twin

Regards,

Max

**Comment by mikedebian** [comment]
Hello Max.

Yes, I did not have X running in my test. I often use console only, and found twin to be perfect for this usecase. I just want to say that I find it to be my favourite console application and I hope that the developement goes well.

You are correct that Links tries to use /dev/tty[0-6], and thus it would not work (according to what I understood of what you wrote).

Links in pure text mode works excellent.

**Comment by mikedebian** [comment]
[links-strace.txt](https://github.com/cosmos72/twin/files/1135470/links-strace.txt)

Here is strace of links -g under twin (no X)

**Comment by cosmos72** [comment]
The console screenshot really looks good :)
Out of curiosity, what did you use to actually *take* the screenshot?

I checked your links-strace.txt and the smoking gun is:

ioctl(0, VIDIOC_RESERVED or VT_GETMODE, 0x55c11d5f48f0) = -1 ENOTTY (Inappropriate ioctl for device)
ioctl(1, VIDIOC_RESERVED or VT_GETMODE, 0x55c11d5f48f0) = -1 ENOTTY (Inappropriate ioctl for device)

i.e. links uses ioctl(VT_GETMODE), a ioctl only supported by Linux console, on standard input and standard output, which are connected to one of /dev/pts/* virtual terminals, that obviously do not support such ioctl.

If you have read/write access to the console as user mxl, you could try some dirty hack like:
  links -g < /dev/tty0
or
  links -g < /dev/console
(no guarantee it works, but it's worth a try)

**Comment by mikedebian** [comment]
I did it through a graphical environment. Started a terminal as fullscreen under X, from which I ran twin. Then I took a screenshot using the desktop screenshot tool.

When I tried to pass links over to another tty, it would not work, nor with the /dev/console.
However, something DOES happen at the other tty as it moves my username to the bottom of the screen and the cursor blinks behind it, instead of in front. Pressing the famous "any key" results in that the cursor moves in front of the username again though. (I do have r/w access). It's hard to explain and harder to take a screenshot of it, sorry.

I could just use another tty and start links from there, but I am so much in love with twin that I am in the process of setting it up as a console environment from which applications will autostart, use it for ssh sessions. I like CLI more than a full blow GUI to be honest.

Also, if it is possible.. 

A custom drop down menu next to File > Window > Custom menu would be nice to set up scripts in. 
I know I can do this in a directory and setup a midnight commander instance to cd into a scripts folder, however... twin could use something like this in my opinion.

Again, thank you for twin. It is absolutely amazing.

**Comment by cosmos72** [comment]
Hmm... it was just a blind try, I am not suprised.

Still, since there was at least **some** effect, it could be interesting to see what happens if you run twin in Linux console, let's say /dev/tty1, then run "strace links -g < /dev/tty1" inside twin (replacing tty1 with the actual tty you are using) and post strace output here.

The custom menu is already possibile - actually there are two options:
1) customize twin menu from $HOME/.twinrc (read system.twinrc in the source tree for an explanation of what can be placed there and how to write it) - the look & feel will be the same as existing twin menus.
2) use twdialog included with twin source tree (it is a reimplementation of plain "dialog" with similar look & feel), for example starting from $HOME/.twinrc a shell script that runs twdialog in a pseudo-tty and performs actions depending on twdialog output

P.S. for history, I actually wrote twin in 1992 under DOS. It exists more or less in its current form (Unix terminal multiplexer) since 1999... it seems I did not try to advertise it enough - nor to include it in some Linux distribution

**Comment by mikedebian** [comment]
[strace.output.txt](https://github.com/cosmos72/twin/files/1137219/strace.output.txt)

Here is strace output as requested and I will look in to the other suggestions! 
I should note that w3m supports images, but when trying to load images through for example google.com under twin, the images get splattered all over the screen and not in their respective places (outside of terminal.) Moving the twin-terminal around clears the images, however it will not show in the terminal. I guess that's a seperate issue.

1992, that's a long time. I'm really glad you've kept working on it. I found twin by accident after trying to find the perfect terminal for the console. It's a shame it hasn't been adopted widely as a replacement for getty. It does in my opinion, crush the competition.

Do you have a donation page? If you don't and use paypal, pm me your paypal so I can send donation your way. I really want to.

**Comment by cosmos72** [comment]
In your new strace log, links went further :) Unluckily it stopped due to
ioctl(0, VT_SETMODE, 0x55f983f128f8)    = -1 EPERM (Operation not permitted)
I do not have an immediate solution for this, I will do some tests myself in the next days.

About the w3m images that do not respect  terminal boundaries: it's already surprising they get displayed at all, since twin does not support images. No womder they garble the screen...

Thanks a lot for the offer to donate!
I have paypal and it's already set-up to accept donations, even though for a different program (stmx, a library I wrote in Common Lisp).
If you want, you can donate through it - you just need to click on the "donate" button on stmx.org (my email address for donations is massimiliano.ghilardi+whatever-you-want@gmail.com
Otherwise I can set up a donation channel specific for twin, but I will not have time for that before Wednesday.

Regards,

Max

**Comment by mikedebian** [comment]
Showing images and videos over twin works perfectly, if done in fullscreen. It also works in windowed mode, but will need to refresh the screen afterwards. I think it uses a framebuffer overlay over twin or something. I have absolutely no idea.

And no problem about the donation. Just a very small thank you for an amazing piece of software, with no strings attached. I hope you'll continue to work on twin, though :)

Anyway, back to the bug! I hope it's possible to fix it.

**Comment by mikedebian** [comment]
Also, where can I find info about TWDM? It installs itself under /usr/local/sbin in opensuse, however that is not my $PATH. On opensuse, it is /usr/sbin. Specify when configure perhaps? I do not know what other distributions use, but I'm assuming it is in fact /usr/local/sbin, so I don't know if it is a bug, or if it is intended. If so, I will add it to my path, or link it.

**Comment by cosmos72** [comment]
Donation received. Thanks a lot!

The install path is intended: all twin executables, libraries and configuration files are installed under /usr/local to avoid messing with "official" software from the Linux/Unix distribution. You can move twdm to another location if you want, it will still work. Or you can even compile and install all twin with --prefix=/usr

As per https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard, twdm should be installed either in /usr/local/sbin (default) or /usr/sbin or /sbin because only root is supposed to run it.

I never wrote a systemd service for twdm, if you want to share what you did I'd be grateful :)

**Comment by mikedebian** [comment]
Well, I'm a tinkerer, not a developer. However, this is what I did.

```
/etc/systemd/system/twdm.service
```
```
[Unit]
Description=twin on tty1

[Service]
Type=simple
ExecStart=/usr/local/sbin/twdm
ExecStop=/bin/kill -HUP ${MAINPID}
StandardInput=tty
StandardOutput=tty
TTYPath=/dev/tty1
Restart=always
RestartSec=2

[Install]
WantedBy=getty.target
```
```
# systemctl disable getty@tty1
# systemctl enable twdm.service
# systemctl stop getty@tty1.
# systemctl daemon-reload
```
It should now be on tty1. Be careful though to make sure everything is correct. This is my setup.
If it's not, do not reboot and go over the details again.

**Comment by cosmos72** [comment]
I am a newbie with systemd, but it looks good to me :) thanks!

It's normal that "suspend", "detach" and "quit" reload twdm...
When twin is started from twdm, only "quit" is actually useful: I am quite sure "suspend" and "detach"  leave a twin server running (or suspended) in background with no display... you will need twattach or twdisplay to reconnect it

**Comment by mikedebian** [comment]
Also another problem is a game called Return, which is a framebuffer game. It seems that it is not allowed to capture any key input. It just displays a static image over twin. Clicking the mouse or moving the keys shows the TWIN window flashing in the background.

Can this be related to links having issues as well? In how TWIN handles the framebuffer?

**Comment by cosmos72** [comment]
Here is the result of my troubleshooting various combinations of links, twin, and framebuffer:
1. links -g (on /dev/tty1): FAIL "fb: cannot open /dev/fb0"
2. su; cd /usr/bin; chown root:video links; chmod g+s links; exit; links -g (on /dev/tty1): SUCCESS
3. links -g <> /dev/tty1 (still sgid video from step 2, this time started inside twin running on /dev/tty1): FAIL "fb: Could not set VT mode."
4. su; cd /usr/bin; chmod y+s links; exit; links -g (on /dev/tty1): SUCCESS
5. links -g (still suid root from step 4, this time started inside twin running on /dev/tty1): FAIL "fb: Could not get VT mode."
6. links -g <> /dev/tty1 (still suid root from step 4, again started inside twin running on /dev/tty1): SUCCESS

So is seems at least **one** of the following conditions must be satisfied
* links controlling tty is one of /dev/tty* **and** it is sgid video (to get read/write access to /dev/fb0) 
* links standard input or output is one of /dev/tty* **and** is sgid root (to bypass kernel check of the controlling tty)

Given such results, the only way I found to have "links -g" working **inside** twin+framebuffer is the ugly one (i.e. dangerous): make "links" setuid root and launch it as "links -g <> /dev/ttyX" where /dev/ttyX is the framebuffer console where twin is running.

Other framebuffer programs will probably have similar problems - if you tell where I can find "Return" game I will have a look too.

**Comment by mikedebian** [comment]
Here it is

It is called ["The Return"](http://betteros.org/games/download.php)

**Comment by cosmos72** [comment]
I just checked, getting ["The Return"](http://betteros.org/games/download.php) to run is even more tricky than links:
it needs access to both /dev/fb0 (0660 root:video on my box) and to /dev/input/event0 (0660 root:input on my box).
So I had to make it setuid root (su; chown root:root Return; chmod u+s Return; exit) - with that, it worked  both from the console (/dev/tty1) and from twin running on the console.
An alternative solution could be changing the group of /dev/fb0 and /dev/input/event0 to the same group id, then mark the executable setgid with that group id, but such changes would probably be lost on reboot.
Also, I had to follow the README and switch to /dev/tty8 after starting the game, to avoid sending keypresses to the console (or to twin)

In summary, using the framebuffer feels a lot like using DOS: to run programs successfully, you have to tinker with OS configuration all the time. And of course, adding twin to the scenario complicates it further...

**Comment by mikedebian** [comment]
> In summary, using the framebuffer feels a lot like using DOS: to run programs successfully, you have to tinker with OS configuration all the time. And of course throwing twin in the middle, further complicates the scenario...

Ah yes :) But then again if you did throw twin in the middle some time in the future, a long with an emulated framebuffer device that would lock application within a window that twin could resize, one would never need X or Wayland for any GUI intents and purposes. Twin, the desktop environment for the console, for those who love the console.
Hey, one can dream. I'll be watching this repo for updates. 

But, back to business...
I'm guessing this bug is nofix, as links could handle it differently (for example, using fbdev2 or drm, which works with twin) and it is not twins problem.

**Comment by cosmos72** [comment]
Yes, basically it's a nofix - there is nothing twin can do in this case.

Linux framebuffer is not designed for sharing among multiple simultaneous processes, and the Linux kernel has two security checks (to prevent misuse) that make even "framebuffer ownership transfer" a bit difficult.

About fbdev2 and drm: do you have some examples that work inside twin? I am curious :)

**Comment by mikedebian** [comment]
Sure, for example with drm, you can for example browse reddit with the application rtv.

If you want to view images, you have to have the w3mimgdisplay package installed, and if you want to watch videos/gifs, you use mpv.

This is all configured in the .mailcap file as such

```
video/x-youtube; mpv -vo drm -quiet '%s'; needsterminal
video/*; mpv -vo drm -quiet '%s'; needsterminal
```

The above is for video.

This means I can browse reddit in RTV using the TWIN-TERMINAL within TWIN, and watch streaming movies and images within TWIN, using RTV to load them.

Or, if you just want to load a standalone video file

```
mpv -vo drm -quiet youtubelinkhere.com.or.videofile.extension.mp4
```

You can also watch images as I said with w3mimgdisplay

On opensuse, that is part of the w3m-inline-image package.

The standalone command to view an image within TWIN is

```
w3m -o 'ext_image_viewer=off' image.extension
```

Sorry about my english.

**Comment by mikedebian** [comment]
In case you are interested, this is how twin handles w3m with images. It's an MP4 video file.

[link here](https://www.dropbox.com/s/1ol9x0aphpgsj3u/w3m%20test.mp4?dl=0)

**Comment by cosmos72** [comment]
I tried w3m with images: I know how it draws pictures, but not how it **finds** on which X11 window it should draw.

I will test rtv and mpv, thanks
